src/zkevm_specs/evm/execution/sdiv_smod.py

```diff
@@ -0,0 +1,109 @@
+from ..instruction import Instruction, Transition
+from ..opcode import Opcode
+from ...util import FQ, RLC
+
+
+def sdiv_smod(instruction: Instruction):
+    opcode = instruction.opcode_lookup(True)
+
+    pop1 = instruction.stack_pop()
+    pop2 = instruction.stack_pop()
+    push = instruction.stack_push()
+
+    (quotient, divisor, remainder, dividend) = __gen_witness(instruction, opcode, pop1, pop2, push)
+    __check_witness(instruction, quotient, divisor, remainder, dividend)
+
+    instruction.step_state_transition_in_same_context(
+        opcode,
+        rw_counter=Transition.delta(3),
+        program_counter=Transition.delta(1),
+        stack_pointer=Transition.delta(1),
+    )
+
+
+def __check_witness(
+    instruction: Instruction, quotient: RLC, divisor: RLC, remainder: RLC, dividend: RLC
+):
+    quotient_is_neg = instruction.word_is_neg(quotient)
+    divisor_is_neg = instruction.word_is_neg(divisor)
+    remainder_is_neg = instruction.word_is_neg(remainder)
+    dividend_is_neg = instruction.word_is_neg(dividend)
+
+    quotient_is_non_zero = 1 - instruction.word_is_zero(quotient)
+    divisor_is_non_zero = 1 - instruction.word_is_zero(divisor)
+    remainder_is_non_zero = 1 - instruction.word_is_zero(remainder)
+
+    quotient_abs = instruction.abs_word(quotient)
+    divisor_abs = instruction.abs_word(divisor)
+    remainder_abs = instruction.abs_word(remainder)
+    dividend_abs = instruction.abs_word(dividend)
```

Comment:
We can simplify the `abs_word` a bit. You can construct the `abs_word` cells first and then prove `abs_word + word == 1 << 256` when `word` is negative. Then the constraint becomes
```
(abs_word_lo + word_lo - 1 << 128) * word_is_neg == 0
(abs_word_hi + word_hi + 1 - 1 << 128) * word_is_neg == 0
```
and no is_zero gadgets are needed.

Reply:
Since there is a special case that `abs(-(1 << 255))` should be `-(1 << 255)` (`1 << 255` is overflow). I updated the constraints as below:
```
# Check if `x == -(1 << 255)`, since `abs(-(1 << 255))` should be equal to `-(1 << 255)`.
is_overflow = self.is_equal(FQ(x.le_bytes[31]), FQ(128)) * self.is_zero(self.sum(x.le_bytes[:31]))
 
# Both `x_lo` and `x_abs_lo` are zero if overflow.
self.constrain_zero((x_abs_lo + x_lo - (1 << 128) * (1 - is_overflow)) * is_neg)

# Both `x_hi` and `x_abs_hi` are `1 << 127` if overflow.
self.constrain_zero((x_abs_hi + x_hi + 1 - is_overflow - (1 << 128)) * is_neg)
```